From 458e94f966785f2aa65b2e80b9136d2e043ae39b Mon Sep 17 00:00:00 2001
From: Liang Li <liang.li@windriver.com>
Date: Tue, 15 Jan 2013 14:52:30 +0800
Subject: [PATCH] wireless/cfg80211: do err handling kfree without holding
 spin-lock

The error handling code in cfg80211_mlme_register_mgmt calling kfree
while holding spin lock wdev->mgmt_registration_lock, then release the
lock right after the kfree.

There is no restriction to the sequence that do kfree before release
spin lock .. and do kfree after release spin lock should be better way
to do error clean up work.

So just minor change to move the kfree after lock release.

Signed-off-by: Lans Zhang <jia.zhang@windriver.com>
Signed-off-by: Liang Li <liang.li@windriver.com>
---
 net/wireless/mlme.c | 7 ++++---
 1 file changed, 4 insertions(+), 3 deletions(-)

diff --git a/net/wireless/mlme.c b/net/wireless/mlme.c
index 7aae329..1a57069 100644
--- a/net/wireless/mlme.c
+++ b/net/wireless/mlme.c
@@ -440,10 +440,8 @@ int cfg80211_mlme_register_mgmt(struct wireless_dev *wdev, u32 snd_portid,
 		}
 	}
 
-	if (err) {
-		kfree(nreg);
+	if (err)
 		goto out;
-	}
 
 	memcpy(nreg->match, match_data, match_len);
 	nreg->match_len = match_len;
@@ -457,6 +455,9 @@ int cfg80211_mlme_register_mgmt(struct wireless_dev *wdev, u32 snd_portid,
  out:
 	spin_unlock_bh(&wdev->mgmt_registrations_lock);
 
+	if (err)
+		kfree(nreg);
+
 	return err;
 }
 
-- 
2.0.2

